openpype/plugins/publish/extract_review.py

```diff
@@ -777,73 +786,53 @@ def fill_sequence_gaps(self, files, staging_dir, start_frame, end_frame):
 
         """
-        start_frame = int(start_frame)
-        end_frame = int(end_frame)
         collections = clique.assemble(files)[0]
         msg = "Multiple collections {} found.".format(collections)
         assert len(collections) == 1, msg
         col = collections[0]
 
-        # do nothing if no gap is found in input range
-        not_gap = True
-        for fr in range(start_frame, end_frame + 1):
-            if fr not in col.indexes:
-                not_gap = False
-
-        if not_gap:
+        start_frame = int(start_frame)
+        end_frame = int(end_frame)
+        missing_indexes = [
+            frame for frame in range(start_frame, end_frame + 1)
+            if frame not in col.indexes
+        ]
+        if not missing_indexes:
             return []
 
-        holes = col.holes()
-
-        # generate ideal sequence
-        complete_col = clique.assemble(
-            [("{}{:0" + str(col.padding) + "d}{}").format(
-                col.head, f, col.tail
-            ) for f in range(start_frame, end_frame)]
-        )[0][0]  # type: clique.Collection
-
-        new_files = {}
-        last_existing_file = None
-
-        for idx in holes.indexes:
-            # get previous existing file
-            test_file = os.path.normpath(os.path.join(
-                staging_dir,
-                ("{}{:0" + str(complete_col.padding) + "d}{}").format(
-                    complete_col.head, idx - 1, complete_col.tail)))
-            if os.path.isfile(test_file):
-                new_files[idx] = test_file
-                last_existing_file = test_file
-            else:
-                if not last_existing_file:
-                    # previous file is not found (sequence has a hole
-                    # at the beginning. Use first available frame
-                    # there is.
-                    try:
-                        last_existing_file = list(col)[0]
-                    except IndexError:
-                        # empty collection?
-                        raise AssertionError(
-                            "Invalid sequence collected")
-                new_files[idx] = os.path.normpath(
-                    os.path.join(staging_dir, last_existing_file))
-
-        files_to_clean = []
-        if new_files:
-            # so now new files are dict with missing frame as a key and
-            # existing file as a value.
-            for frame, file in new_files.items():
-                self.log.info(
-                    "Filling gap {} with {}".format(frame, file))
+        # Both collection frame indexes and the missing indexes are sorted
+        # so we can keep track of the nearest existing frame with an iterator
+        existing_indices_iter = iter(col.indexes)
+        prev_idx = next(existing_indices_iter, None)
+        next_idx = next(existing_indices_iter, prev_idx)
+        if prev_idx is None:
+            raise RuntimeError("No indices in input collection. This is a bug")
```

Comment:
Please us `KnownPublishError` from `openpype.pipeline.publish`.

Reply:
It actually isn't a known error. This should technically NEVER happen. Since it wouldn't have an assembled collection to begin with earlier in the code if there would've been no "indices". This really would be an odd code error/bug if it would occur.

Reply:
If you raise error before other possible error then it's known error. If it should never happen then the check is redundant and should be removed...

Note: I would simplify it to this:
```
        first_frame = col.indexes[0]
        prev_frame = None
        hole_frame_to_nearest = {}
        # First prepare which hole is filled with frame
        for frame in range(int(frame_start), int(frame_end) + 1):
            if frame in col.indexes:
                prev_frame = frame
            elif prev_frame is None:
                hole_frame_to_nearest[frame] = first_frame
            else:
                hole_frame_to_nearest[frame] = prev_frame
        
        # Calculate paths
        holes_to_nearest = {}
        for hole_frame, src_frame in hole_frame_to_nearest.items():
            src_fpath = os.path.join(staging_dir, col_format % src_frame)
            hole_fpath = os.path.join(staging_dir, col_format % hole_frame)
            holes_to_nearest[hole_fpath] = nearest_fpath
```

Reply:
I agree that this is indeed a known error case or there'sno point in raising it. By definition if you are catching it and raising it, you know that is can happen and should use the `KnownPublishError` so we can make any know errors less scary for the user.

Reply:
@BigRoy could we address this? we'll hopefully get to some proper testing on more complicated data soon.